src/models/Logo.php

```diff
@@ -156,28 +224,133 @@ private static function logoFromRow(Map<string, string> $row): Logo {
     bool $used,
     bool $enabled,
     bool $protected,
+    bool $custom,
     string $name,
     string $logo,
-  ): Awaitable<int> {
+  ): Awaitable<Logo> {
     $db = await self::genDb();
 
     // Create category
     await $db->queryf(
-      'INSERT INTO logos (used, enabled, protected, name, logo) VALUES (%d, %d, %d, %s, %s)',
+      'INSERT INTO logos (used, enabled, protected, custom, name, logo) VALUES (%d, %d, %d, %d, %s, %s)',
       $used ? 1 : 0,
       $enabled ? 1 : 0,
       $protected ? 1 : 0,
+      $custom ? 1 : 0,
       $name,
       $logo,
     );
 
     // Return newly created logo_id
-    $result = await $db->queryf(
-      'SELECT id FROM logos WHERE logo = %s LIMIT 1',
-      $name,
-    );
+    $result =
+      await $db->queryf('SELECT * FROM logos WHERE logo = %s LIMIT 1', $logo);
 
     invariant($result->numRows() === 1, 'Expected exactly one result');
-    return intval($result->mapRows()[0]['id']);
+    self::invalidateMCRecords();
+    return self::logoFromRow($result->mapRows()[0]);
+  }
+
+  // Create custom logo
+  public static async function genCreateCustom(
+    string $base64_data,
+  ): Awaitable<?Logo> {
+    // Check image size
+    $image_size_bytes = strlen($base64_data) * self::BASE64_BYTES_PER_CHAR;
+    if ($image_size_bytes > self::MAX_CUSTOM_LOGO_SIZE_BYTES) {
+      error_log(
+        'Logo file base64 not less than '.
+        (self::MAX_CUSTOM_LOGO_SIZE_BYTES / 1000).
+        ' kB, was '.
+        ($image_size_bytes / 1000).
+        ' kB',
+      );
+      return null;
+    }
+
+    // Get image properties and verify mimetype
+    $base64_data = str_replace(' ', '+', $base64_data);
+    $binary_data = base64_decode(str_replace(' ', '+', $base64_data));
+    $image_info = getimagesizefromstring($binary_data);
+
+    $mimetype = $image_info[2];
+
+    if (!array_key_exists($mimetype, self::$CUSTOM_LOGO_TYPES)) {
+      error_log("Image type '$mimetype' not allowed");
+      return null;
+    }
+
+    $binary_data =
+      self::resizeCustomLogo($binary_data, $image_info[1], $image_info[0]);
+
+    $type_extension = self::DEFAULT_LOGO_TYPE;
+
+    $filename = 'custom-'.time().'-'.md5($base64_data).'.'.$type_extension;
```

Comment:
Easily changeable, but md5 is already being used on Attachment uploads.  Maybe this is something that should be changed across the board?